### DISCUSSION TITLE
Stricter ChangeHandler type

### DISCUSSION THREAD

**Discussion by ThomasVandenbon** [react-hook-form/react-hook-form#4797] [Q&A]
I'm trying to make sense of the code in [useController.ts](https://github.com/react-hook-form/react-hook-form/blob/master/src/useController.ts#L70) compared to code in [useForm.ts](https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.ts#L593). 

Both are making use of the `ChangeHandler`-type, which sadly has an `event`-parameter of type `any`.
```ts
export type ChangeHandler = (event: any) => Promise<void | boolean>;
```
I'm trying to figure out what values would be accepted for this `event`-parameter, in other words I'm trying to make the type more strict.

The code in `useController` (and `getControllerValue`) leads me to believe that any value (`object`, `string`, `number`, ...) would work and be converted to a type that resembles an event.
This type that resembles an event seems to look something like:
```ts
type EventCreatedInUseController = {
  target: {
    value: XXXX, // A type derived from the provided event-parameter
    name: InternalFieldName,
  },
  type: 'change',
}
```

However, the code in `useForm` seems to indicate that the value must at least have a `type` and a `target`, where the `target` contains a `value` and a `type`.
Meaning the event parameter should look something like this:
```ts
type EventParameter = {
  type: 'blur' | unknown, // Should probably be 'blur' | 'change'
  target: {
    value: XXXX, // Could be anything?
    type: string // Could be anything?
  }
}
```

Am I right that there is a difference between these 2 uses of `ChangeHandler`?
Is this difference on purpose?
Would it be possible to make the type more strict (potentially by splitting `ChangeHandler` into 2 separate types)?

**Comment by bluebill1049** [comment]
I am open for suggestion and PR for that.

**Comment by wibed** [comment]
i think this has been dealt with and can be closed:
https://github.com/react-hook-form/react-hook-form/issues/6018
